File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys
input = sys.stdin.readline

n,k = map(int, input().split())
put = [[1, 1]]
bag = [[0 for _ in range(k + 1)] for _ in range(n + 1)]

for _ in range(n):
    put.append(list(map(int, input().split())))

for i in range(1, n + 1):
    for j in range(1, k + 1):
        kg = put[i][0]
        val = put[i][1]

        if j < kg:
            bag[i][j] = bag[i - 1][j]
        else:
            bag[i][j] = max(val + bag[i - 1][j - kg], bag[i - 1][j])

print(bag[n][k])

File: acmicpc/#12865 평범한 배낭/평범한 배낭.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [N, K] = input[0].split(' ').map(Number);
const items = input.slice(1).map(line => line.split(' ').map(Number));

const dp = Array.from(Array(N + 1), () => Array(K + 1).fill(0));

for (let i = 1; i <= N; i++) {
    const [weight, value] = items[i - 1];

    for (let j = 1; j <= K; j++) {
        if (j < weight) {
            dp[i][j] = dp[i - 1][j];
        } else {
            dp[i][j] = Math.max(value + dp[i - 1][j - weight], dp[i - 1][j]);
        }
    }
}

console.log(dp[N][K]);
